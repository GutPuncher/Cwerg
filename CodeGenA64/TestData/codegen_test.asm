
.mem COUNTER 4 RW
.data 4 [0]

.fun TestMoveImmediates NORMAL [] = [A64]
  .reg U32 [$r0_U32 $r1_U32 $r2_U32 rem div]
  .reg S32 [$r0_S32]
  .reg U8 [$r0_U8]
  .reg A64 [addr]
.bbl start
  .stk buffer 1 16
    conv $r0_U32@w10 $r0_U8@w10
    mov $r0_S32@w11 -100:S32
    mov $r0_U32@w10 525032:U32
    mov addr@x20 0x1234567890123456:A64
    lea addr@x20 = buffer 0:U32
    lea.mem addr@x20 COUNTER 100:S32
    lea.mem addr@x20 COUNTER $r0_U32@w10
    mov $r1_U32@w11 $r0_U32@w10
    mov $r0_U32@w10 10
    div rem@w5 $r1_U32@w11 $r0_U32@w10
    mov $r0_U32@w10 10
    mul rem@w5 rem@w5 $r0_U32@w10
    sub rem@w5 $r1_U32@w11 rem@w5
    mov $r0_U32@w10 10
    bne  $r0_U32@w10 148932 skip
    mov $r0_U32@w10 148932
    div div@w6 $r1_U32@w11 $r0_U32@w10
    beq div@w6 0 skip

.bbl skip
    poparg base:A64
    ld add:A64@x23  base@x22 0:U32
    ret


.fun TestMoveImmediates_gpr_scratch NORMAL [] = [A64]
  .reg S32 [$r0_S32]
  .reg U8 [$r0_U8]
  .reg A64 [addr]
  .reg F32 [flt]
  .reg F64 [dbl]
   .stk buffer 16 16
.bbl start
     copysign flt@s1 flt@s1 0:F32   # abs
     copysign dbl@d2 dbl@d2 0:F64   # abs
     st.stk buffer 0:U32 dbl
     ld.stk flt buffer 0:U32
     beq dbl@d2 0.0 skip
     copysign flt@s1 flt@s1 flt@s1
     copysign dbl@d2 dbl@d2 dbl@d2
.bbl skip
     ret



