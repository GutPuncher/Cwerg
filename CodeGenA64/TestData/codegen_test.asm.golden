
INS: conv $r0_U32@w10 $r0_U8@w10  [U32 U8]
PAT: reg:[U32 U8]  imm:[INVALID INVALID]
    and_x_imm x10 x10 0xff

INS: mov $r0_S32@w11 -100  [S32 S32]
PAT: reg:[S32 S32]  imm:[INVALID IMM_SHIFTED_5_20_21_22_NOT]
    movn_x_imm x11 0x63

INS: mov $r0_U32@w10 525032  [U32 U32]
PAT: reg:[U32 U32]  imm:[INVALID IMM_POS_32]
    movz_x_imm x10 0x2e8
    movk_x_imm x10 0x80000

INS: mov addr@x20 1311768467284833366  [A64 A64]
PAT: reg:[A64 A64]  imm:[INVALID ANY]
    movz_x_imm x20 0x3456
    movk_x_imm x20 0x90120000
    movk_x_imm x20 0x567800000000
    movk_x_imm x20 0x1234000000000000

INS: lea.stk addr@x20 buffer 0  [A64 _ U32]
PAT: reg:[A64 INVALID U32]  imm:[INVALID INVALID pos_stk_combo_shifted_10_21_22]
    add_x_imm x20 sp 0x0

INS: mov $r1_U32@w11 $r0_U32@w10  [U32 U32]
PAT: reg:[U32 U32]  imm:[INVALID INVALID]
    orr_x_reg x11 xzr x10 lsl 0

INS: mov $r0_U32@w10 10  [U32 U32]
PAT: reg:[U32 U32]  imm:[INVALID IMM_SHIFTED_5_20_21_22]
    movz_x_imm x10 0xa

INS: div rem@w5 $r1_U32@w11 $r0_U32@w10  [U32 U32 U32]
PAT: reg:[U32 U32 U32]  imm:[INVALID INVALID INVALID]
    udiv_w w5 w11 w10

INS: mov $r0_U32@w10 10  [U32 U32]
PAT: reg:[U32 U32]  imm:[INVALID IMM_SHIFTED_5_20_21_22]
    movz_x_imm x10 0xa

INS: mul rem@w5 rem@w5 $r0_U32@w10  [U32 U32 U32]
PAT: reg:[U32 U32 U32]  imm:[INVALID INVALID INVALID]
    madd_w w5 w5 w10 wzr

INS: sub rem@w5 $r1_U32@w11 rem@w5  [U32 U32 U32]
PAT: reg:[U32 U32 U32]  imm:[INVALID INVALID INVALID]
    sub_w_reg w5 w11 w5 lsl 0

INS: mov $r0_U32@w10 10  [U32 U32]
PAT: reg:[U32 U32]  imm:[INVALID IMM_SHIFTED_5_20_21_22]
    movz_x_imm x10 0xa

INS: bne $r0_U32@w10 148932 skip  [U32 U32 _]
    mismatches: 2

INS: mov $r0_U32@w10 148932  [U32 U32]
PAT: reg:[U32 U32]  imm:[INVALID IMM_POS_32]
    movz_x_imm x10 0x45c4
    movk_x_imm x10 0x20000

INS: div div@w6 $r1_U32@w11 $r0_U32@w10  [U32 U32 U32]
PAT: reg:[U32 U32 U32]  imm:[INVALID INVALID INVALID]
    udiv_w w6 w11 w10

INS: beq div@w6 0 skip  [U32 U32 _]
PAT: reg:[U32 U32 INVALID]  imm:[INVALID IMM_SHIFTED_10_21_22 INVALID]
    subs_w_imm wzr w6 0x0
    b_eq expr:condbr19:skip

INS: poparg base  [A64]
    MATCH_IMPOSSIBLE

INS: ld add base 0  [A64 A64 U32]
    mismatches: 4

INS: ret  []
    SPECIAL
